cmd/influx-tools/importer/importer.go

```diff
@@ -0,0 +1,243 @@
+			return false
+		}
+		i.store = store
+	}
+
+	if !buildTsi {
+		i.seriesBuf = make([]byte, 0, 2048)
+	}
+	return i
+}
+
+func (i *importer) Open() error {
+	if i.replace {
+		return i.store.Open()
+	}
+	return nil
+}
+
+func (i *importer) Close() error {
+	el := errlist.NewErrorList()
+	if i.sh != nil {
+		el.Add(i.CloseShardGroup())
+	}
+	if i.replace {
+		el.Add(i.store.Close())
+	}
+	return el.Err()
+}
+
+func (i *importer) CreateDatabase(rp *meta.RetentionPolicySpec) error {
+	var rpi *meta.RetentionPolicyInfo
+	dbInfo := i.MetaClient.Database(i.db)
+	if dbInfo == nil {
+		return i.createDatabaseWithRetentionPolicy(rp)
+	}
+
+	rpi, err := i.MetaClient.RetentionPolicy(i.db, rp.Name)
+	if err != nil {
+		return err
+	}
+
+	updateRp := (rpi != nil) && i.replace &&
+		((rp.Duration != nil && rpi.Duration != *rp.Duration) || (rpi.ShardGroupDuration != rp.ShardGroupDuration))
+	if updateRp {
+		err = i.updateRetentionPolicy(rpi, rp)
+		if err != nil {
+			return err
+		}
+	} else {
+		_, err = i.MetaClient.CreateRetentionPolicy(i.db, rp, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	return i.createDatabaseWithRetentionPolicy(rp)
+}
+
+func (i *importer) updateRetentionPolicy(oldRpi *meta.RetentionPolicyInfo, newRp *meta.RetentionPolicySpec) error {
+	for _, shardGroup := range oldRpi.ShardGroups {
+		err := i.removeShardGroup(&shardGroup)
+		if err != nil {
+			return err
+		}
+	}
+
+	rpu := &meta.RetentionPolicyUpdate{Name: &newRp.Name, Duration: newRp.Duration, ShardGroupDuration: &newRp.ShardGroupDuration}
+
+	return i.MetaClient.UpdateRetentionPolicy(i.db, newRp.Name, rpu, false)
+}
+
+func (i *importer) createDatabaseWithRetentionPolicy(rp *meta.RetentionPolicySpec) error {
+	var err error
+	var dbInfo *meta.DatabaseInfo
+	if len(rp.Name) == 0 {
+		dbInfo, err = i.MetaClient.CreateDatabase(i.db)
+	} else {
+		dbInfo, err = i.MetaClient.CreateDatabaseWithRetentionPolicy(i.db, rp)
+	}
+	if err != nil {
+		return err
+	}
+	i.rpi = dbInfo.RetentionPolicy(rp.Name)
+	return nil
+}
+
+func (i *importer) StartShardGroup(start int64, end int64) error {
+	sgi := i.rpi.ShardGroupByTimestamp(time.Unix(0, start))
+	if sgi != nil {
+		if i.replace {
+			err := i.removeShardGroup(sgi)
+			if err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("shard already exists for start time %v", start)
+		}
+	}
+
+	sgi, err := i.MetaClient.CreateShardGroup(i.db, i.rpi.Name, time.Unix(0, start))
+	if err != nil {
+		return err
+	}
+
+	shardPath := filepath.Join(i.dataDir, i.db, i.rpi.Name)
+	if err = os.MkdirAll(filepath.Join(shardPath, strconv.Itoa(int(sgi.ID))), 0777); err != nil {
+		return err
+	}
+
+	i.sh = newShardWriter(sgi.ID, shardPath)
+
+	i.startSeriesFile()
+	return nil
+}
+
+func (i *importer) removeShardGroup(sgi *meta.ShardGroupInfo) error {
+	err := i.MetaClient.DeleteShardGroup(i.db, i.rpi.Name, sgi.ID)
+	if err != nil {
+		return err
+	}
+
+	i.store.EngineOptions.ShardFilter = func(_, _ string, id uint64) bool {
```

Comment:
You will need to remove the use of `TSDBStore#Open` and `DeleteShard` APIs as this could cause corruption / conflicts given you also open the series file and indexes for writing. 

My inclination is you will need to manage deletion of the shards in the import tool.